pkg/sbproxy/reconcile/reconciler.go

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2018 The Service Manager Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reconcile
+
+import (
+	"context"
+
+	"github.com/Peripli/service-broker-proxy/pkg/sbproxy/notifications"
+	"github.com/Peripli/service-broker-proxy/pkg/sbproxy/notifications/handlers"
+
+	"github.com/Peripli/service-broker-proxy/pkg/platform"
+	"github.com/Peripli/service-broker-proxy/pkg/sm"
+	"github.com/Peripli/service-manager/pkg/log"
+	"github.com/Peripli/service-manager/pkg/types"
+	"github.com/gofrs/uuid"
+	cache "github.com/patrickmn/go-cache"
+	"github.com/pkg/errors"
+)
+
+const smBrokersStats = "sm_brokers"
+
+// for stubbing in unit tests
+type executor interface {
+	apply(*types.Notification)
+	resync()
+}
+
+// Reconciler takes care of propagating broker and visibility changes to the platform.
+// It reconciles the state of the proxy brokers and visibilities
+// in the platform to match the desired state provided by the Service Manager.
+// TODO if the reg credentials are changed (the ones under cf.reg) we need to update the already registered brokers
+type Reconciler struct {
+	Options        *Settings
+	PlatformClient platform.Client
+	SMClient       sm.Client
+	ProxyPath      string
+	Cache          *cache.Cache
+	GlobalContext  context.Context
+
+	notificationConsumer *notifications.Consumer
+	exec                 executor
+}
+
+type resyncJob struct {
+	*Reconciler
+
+	resyncContext context.Context
+	stats         map[string]interface{}
+}
+
+// Process resync and notification messages sequentially in one goroutine
+// to avoid concurrent changes in the platform
+func (r *Reconciler) Process(messages <-chan *notifications.Message) {
+	exec := r.exec
+	if exec == nil {
+		exec = r
+	}
+
+	for {
+		select {
+		case <-r.GlobalContext.Done():
+			log.C(r.GlobalContext).Info("Context cancelled. Terminating reconciler.")
+			return
+		case m, ok := <-messages:
+			if !ok {
+				log.C(r.GlobalContext).Info("Messages channel closed. Terminating reconciler.")
+				return
+			}
+			log.C(r.GlobalContext).Debugf("Reconciler received message %+v", m)
+			if m.Resync {
+				// discard any pending change notifications as we will do a full resync
+				drain(messages)
+				exec.resync()
```

Comment:
What if the context is canceled when a SIGTERM is received? Assuming we pass the context down only sm client uses the context when making requests. CF client makes calls that don't use context, so these will block until CF sends SIGKILL after 10 seconds. 
I assume this wouldn't be a big problem given that we do a full resync on startup, right?